### PR TITLE
gh-109653: Improve import time of importlib.metadata / email.utils

### PR DESCRIPTION
**Pull request by hauntsaninja**
My criterion for delayed imports is that they're only worth it if the majority of users of the module would benefit from it, otherwise you're just moving latency around unpredictably.

mktime_tz is not used anywhere in the standard library and grep.app indicates it's not got much use in the ecosystem either.

Distribution.files is not nearly as widely used as other importlib.metadata APIs, so we defer the csv import.

And it's pretty easy for Python programs to not ever need calendar and csv.

Before:
```
λ hyperfine -w 8 './python -c "import importlib.metadata"'
Benchmark 1: ./python -c "import importlib.metadata"
  Time (mean ± σ):      65.1 ms ±   0.5 ms    [User: 55.3 ms, System: 9.8 ms]
  Range (min … max):    64.4 ms …  66.4 ms    44 runs
```

After:
```
λ hyperfine -w 8 './python -c "import importlib.metadata"'
Benchmark 1: ./python -c "import importlib.metadata"
  Time (mean ± σ):      62.0 ms ±   0.3 ms    [User: 52.5 ms, System: 9.6 ms]
  Range (min … max):    61.3 ms …  62.8 ms    46 runs
```

for about a 3ms saving with warm disk cache, maybe 7-11ms with cold disk cache.

* Issue: gh-109653

### CODE REVIEW

**Review by jaraco — COMMENTED** [review]
Thanks for working on this change. I appreciate the effort.

I'm a little concerned for two reasons.

First, this change implies that best practice is not to put imports at the top of a module, conflicting with my understanding that imports should generally be placed at the top of the module. If there's a systemic problem with eager imports adding undue latency, I'd prefer to see that issue dealt with systemically, rather than piecemeal moving select imports to be late-imported. The approach implied by this change suggests that any Python developer (and especially CPython developers) should be alert to the cost of their imports and aggressively inline imports to be used on demand. I've seen [PEP 690](https://peps.python.org/pep-0690/), but it appears to have been rejected (and without any reason that I could find in the PEP). It seems to me than something like PEP 690 would be better than selectively combing through codebases and analyzing their imports and tweaking them. 

Second, the motivation for this change is lost in the commit - there's nothing in the code or tests that would indicate to a future contributor not to reverse this change for the sake of consistency. At the very least, I'd expect a comment protecting the motivation.

Third, the importlib.metadata part of this change will need to also be contributed to importlib_metadata. That test suite does have performance tests that measure specific performance targets. Perhaps that could be a good place to ensure that the desired behavior is captured.

Given that PEP 690 was rejected, perhaps this kind of change should be enacted through a linter that would automatically scan for imports that are not used until called and automatically move them inline (and thus could also be included in test suites to protect the intention).

I don't have any real objections on this specific change, just a bit of concern at the underlying implications.

**Comment by jaraco** [comment]
> My criterion for delayed imports

Perhaps it would be better if there were a shared, published criteria for delayed imports, either in CPython or for Python libs in general.

**Comment by jaraco** [comment]
> Perhaps that could be a good place to ensure that the desired behavior is captured.

In python/importlib_metadata@1a209e7e41, I started exploring adding a test that would capture the import time expectation (and at least indicate a regression when if the optimization was removed), but when I added the lazy import for csv, it detected no meaningful performance difference:

```
exercises.py:import time: 0:00:00.000001 (+0:00:00, 0%)
```

That indicates to me that the import time is 1 µs before and after the lazy import.

Attempting to understand why the timeit results weren't providing the expected performance gains, I added an assertion and found that the "setup" expression has a different outcome depending on what "test" expression is present:

```
 importlib_metadata feature/test-import-time @ .tox/py/bin/python -s -m timeit --setup 'import sys; assert "csv" not in sys.modules'
100000000 loops, best of 5: 3.9 nsec per loop
 importlib_metadata feature/test-import-time @ .tox/py/bin/python -s -m timeit --setup 'import sys; assert "csv" not in sys.modules' 'import importlib_metadata'
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.12/3.12.1_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/timeit.py", line 330, in main
    number, _ = t.autorange(callback)
                ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.1_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/timeit.py", line 226, in autorange
    time_taken = self.timeit(number)
                 ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.1_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/timeit.py", line 180, in timeit
    timing = self.inner(it, self.timer)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<timeit-src>", line 3, in inner
    import sys; assert "csv" not in sys.modules
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

That's weird, right? I would expect the "setup" to be run exactly once before any test expression was run, so it shouldn't be affected by the contents of the test expression, yet it is.

**Comment by hauntsaninja** [comment]
@jaraco Thank you for taking the time to review. I find `timeit` hard to use to benchmark import time because of Python's caching behaviour. What timeit is doing is something like:
```
for _ in range(r):
    import sys; assert "csv" not in sys.modules
    for _ in range(n):
        import importlib.metadata
```
This explains why you see:
```
# this works
λ python -m timeit -n 1 -r 1 --setup 'import sys; assert "csv" not in sys.modules' 'import importlib.metadata'
1 loop, best of 1: 47.5 msec per loop

# but this fails
λ python -m timeit -n 1 -r 2 --setup 'import sys; assert "csv" not in sys.modules' 'import importlib.metadata'
Traceback (most recent call last):
[...]
AssertionError
```

In general, Python's caching behaviour for imports makes it very easy to benchmark incorrectly when using timeit:
```
λ cat slow.py
import time
time.sleep(1)
λ python -m timeit 'import slow'
1 loop, best of 5: 541 nsec per loop
```

> First, this change implies that best practice is not to put imports at the top of a module

I think this is a reasonable reason to reject this change. Another reasonable request to reject this change is that the benefit is relatively small and as the primary maintainer of this module you'd like a consistent style.

I think of my personal view as pragmatic: `importlib.metadata` is the slowest import in a command line tool I work on. It's unfortunate that my CLI is slower because `importlib.metadata` imports `email.utils` and `email.utils` imports `calendar` only for `mktime_tz`, a function that almost no one uses.

I agree it could be nice to have a more systemic solution. PEP 690 was a global flag to the interpreter, which complicates things, e.g. I think the SC wasn't keen on having libraries double the size of their test matrix. It's also a good point that we could put something in the dev guide about how to think about this.

> Second, the motivation for this change is lost in the commit

Thanks, I'll add a comment :-)

> Third, the importlib.metadata part of this change will need to also be contributed to importlib_metadata

I can make the same patch to `importlib_metadata` :-)

**Review by AlexWaygood — APPROVED** [review]
This LGTM. I agree that these are good candidates for lazy imports, and I agree that it would be nice to improve the import time of `importlib.metadata`, which is a very useful module that unfortunately does have one of the slower import times in the standard library currently. @jaraco should obviously have the final say, though, as the primary maintainer for this module and the `importlib_metadata` backport.

> First, this change implies that best practice is not to put imports at the top of a module, conflicting with my understanding that imports should generally be placed at the top of the module.

While this is a general style recommendation by PEP-8, I think it's worth noting that we already break this in many places across the stdlib, for reasons of pragmatism:

https://github.com/python/cpython/blob/a768e12f094a9b14a9a1680fb50330e1050716c4/Lib/functools.py#L19

https://github.com/python/cpython/blob/a768e12f094a9b14a9a1680fb50330e1050716c4/Lib/email/__init__.py#L28-L37

https://github.com/python/cpython/blob/a768e12f094a9b14a9a1680fb50330e1050716c4/Lib/gettext.py#L351-L353

https://github.com/python/cpython/blob/a768e12f094a9b14a9a1680fb50330e1050716c4/Lib/argparse.py#L141-L147

Also, I'll note that PEP-8 also contains this statement ;)

> A Foolish Consistency is the Hobgoblin of Little Minds

FWIW, I agree that:
- We definitely shouldn't run around the whole standard library making all imports lazy; we should look at each one on a case-by-case basis and check that a specific change is impactful before making it
- We should always add comments to the source code when making changes like this
- It would also be nice if we could speedup Python's `import` statement in general (but I don't think that should block small improvements like this, where they make sense)

**Review by jaraco — APPROVED** [review]

**Comment by jaraco** [comment]
> I can make the same patch to `importlib_metadata`. Note that of the two changes, deferral of `import calendar` in `email._parseaddr` accounts for most of the time saved — this wouldn't apply to `importlib_metadata`.

I'll take this on, since I've got the muscle memory for doing it. I just wanted to flag it for context.

> I prefer using subprocesses or general purpose benchmarking tools

Alex provided a similar suggestion in the linked issue I filed about timeit. I would like for [pytest-perf](/jaraco/pytest-perf) to be able to facilitate the testing of import latency. I'd like for pytest-perf to be one of those general purpose benchmarking tools. I've filed https://github.com/jaraco/pytest-perf/issues/12 to track that possibility.

**Comment by danielhollas** [comment]
>  I prefer using subprocesses or general purpose benchmarking tools

Note that there's also a command line option `-Ximporttime` specifically for disentangling the import time cost. Run e.g. like `python -Ximporttime -c "import importlib.metadata > import_time.out",
producing a text file with the import time of all the modules that end up being imported. (you might want to run it couple times to warm up disc cache, then the results are quite stable in my experience)

You can visualize it with [tuna](https://pypi.org/project/tuna/) to get a nice icicle graph of where the import time goes.

**Comment by hugovk** [comment]
> I've seen [PEP 690](https://peps.python.org/pep-0690/), but it appears to have been rejected (and without any reason that I could find in the PEP).

To see the SC's rejection notice, click through from "Resolution: [Discourse message](https://discuss.python.org/t/pep-690-lazy-imports-again/19661/26)" in the header; essentially they were concerned over producing a split in the community over how imports work.

> You can visualize it with [tuna](https://pypi.org/project/tuna/) to get a nice icicle graph of where the import time goes.

Yes, tuna is a nice tool. https://github.com/hugovk/pypistats/pull/289 is an example of using it to identify and improve slow imports for a CLI.

As it happens, `importlib.metadata` accounts for 27% of the final import time in that Python 3.10 example. Measuring again to zoom in, it accounts for 59% (I think because 3.12 is quicker) with 2.3% for `calendar` and 0.8% for `csv` so this PR is welcome. Thanks all!

**Comment by jaraco** [comment]
Backported in python/importlib_metadata@b4ce0ff90e.

**Pull request merged**
